eurotoken/src/main/java/nl/tudelft/trustchain/eurotoken/community/EuroTokenCommunity.kt

```diff
@@ -71,13 +118,96 @@ class EuroTokenCommunity(
     object MessageId {
         const val GATEWAY_CONNECT = 1
         const val ROLLBACK_REQUEST = 1
+        const val ATTACHMENT = 4
     }
 
     class Factory(
-        private val store: GatewayStore
+        private val store: GatewayStore,
+        private val trustStore : TrustStore,
+        private val context : Context,
     ) : Overlay.Factory<EuroTokenCommunity>(EuroTokenCommunity::class.java) {
         override fun create(): EuroTokenCommunity {
-            return EuroTokenCommunity(store)
+            return EuroTokenCommunity(store, trustStore, context)
         }
     }
+
+    /**
+     * Generate a public key based on the [seed].
+     * @param seed : the seed used to generate the public key.
+     */
+    private fun generatePublicKey(seed : Long) : String {
+        // Initialize Random with seed
+        val random = Random(seed)
+
+        // Generate a random public key of 148 hexadecimal characters
+        val key = random.nextBytes(148)
+        return key.toHex()
+    }
+
+    /**
+     * Generate [numberOfKeys] public keys based on the [seed].
+     * @param numberOfKeys : the number of keys to generate.
+     * @param seed : the seed used to generate the public keys.
+     */
+    private fun generatePublicKeys(numberOfKeys: Int, seed : Long = 1337) : List<String> {
+        val publicKeys = mutableListOf<String>()
+        for (i in 0 until numberOfKeys) {
+            publicKeys.add(generatePublicKey(seed + i))
+        }
+        return publicKeys
+    }
+
+
+    /**
+     * Called after the user has finished a transaction with the other party.
+     * Sends the [num] public keys of latest transaction counterparties to the receiver.
+     * When DEMO mode is enabled, it generates 50 random keys instead.
+     * @param peer : the peer to send the keys to.
+     * @param num : the number of keys to send.
+     */
+    fun sendAddressesOfLastTransactions(peer: Peer, num: Int = 50) {
+        val pref = myContext.getSharedPreferences(EUROTOKEN_SHARED_PREF_NAME, Context.MODE_PRIVATE)
+        val demoModeEnabled = pref.getBoolean(DEMO_MODE_ENABLED, false)
+
+        val addresses : ArrayList<String> = ArrayList()
+        // Add own public key to list of addresses.
+        addresses.add(myPeer.publicKey.keyToBin().toHex())
+        if (demoModeEnabled) {
+            // Generate [num] addresses if in demo mode
+            addresses.addAll(generatePublicKeys(num))
+        } else {
+            // Get all addresses of the last [num] incoming transactions
+            addresses.addAll(transactionRepository.getTransactions(num).map { transaction: Transaction ->
+                transaction.sender.toString()
+            })
+        }
+
+        val payload = TransactionsPayload("1", addresses.joinToString(separator = ",").toByteArray())
```

Comment:
Probably better to use the actual message ID const you defined earlier.

Reply:
Done!